server/etcdserver/api/v3discovery/discovery.go

```diff
@@ -0,0 +1,569 @@
+
+	DialTimeout      time.Duration //--discovery-dial-timeout
+	RequestTimeOut   time.Duration //--discovery-request-timeout
+	KeepAliveTime    time.Duration //--discovery-keepalive-time
+	KeepAliveTimeout time.Duration //--discovery-keepalive-timeout
+
+	InsecureTransport  bool   //--discovery-insecure-transport
+	InsecureSkipVerify bool   //--discovery-insecure-skip-tls-verify
+	CertFile           string //--discovery-cert
+	KeyFile            string //--discovery-key
+	TrustedCAFile      string //--discovery-cacert
+
+	User     string //--discovery-user
+	Password string //--discovery-password
+}
+
+type memberInfo struct {
+	peerId    string
+	peerURL   string
+	createRev int64
+}
+
+type clusterInfo struct {
+	clusterId string
+	members   []memberInfo
+}
+
+// key prefix for each cluster: "/_etcd/registry/<ClassId>".
+func geClusterKeyPrefix(cluster string) string {
+	return path.Join(discoveryPrefix, cluster)
+}
+
+// key format for cluster size: "/_etcd/registry/<ClassId>/_config/size".
+func geClusterSizeKey(cluster string) string {
+	return path.Join(geClusterKeyPrefix(cluster), "_config/size")
+}
+
+// key prefix for each member: "/_etcd/registry/<ClassId>/members".
+func getMemberKeyPrefix(cluster string) string {
+	return path.Join(geClusterKeyPrefix(cluster), "members")
+}
+
+// key format for each member: "/_etcd/registry/<ClassId>/members/<memberId>".
+func getMemberKey(cluster, memberId string) string {
+	return path.Join(getMemberKeyPrefix(cluster), memberId)
+}
+
+// GetCluster will connect to the discovery service at the given url and
+// retrieve a string describing the cluster
+func GetCluster(lg *zap.Logger, dUrl string, cfg *DiscoveryConfig) (cs string, rerr error) {
+	d, err := newDiscovery(lg, dUrl, cfg, 0)
+	if err != nil {
+		return "", err
+	}
+
+	defer d.close()
+	defer func() {
+		if rerr != nil {
+			d.lg.Error(
+				"discovery failed to get cluster",
+				zap.String("cluster", cs),
+				zap.Error(rerr),
+			)
+		} else {
+			d.lg.Info(
+				"discovery got cluster successfully",
+				zap.String("cluster", cs),
+			)
+		}
+	}()
+
+	return d.getCluster()
+}
+
+// JoinCluster will connect to the discovery service at the given url, and
+// register the server represented by the given id and config to the cluster
+func JoinCluster(lg *zap.Logger, durl string, cfg *DiscoveryConfig, id types.ID, config string) (cs string, rerr error) {
+	d, err := newDiscovery(lg, durl, cfg, id)
+	if err != nil {
+		return "", err
+	}
+
+	defer d.close()
+	defer func() {
+		if rerr != nil {
+			d.lg.Error(
+				"discovery failed to join cluster",
+				zap.String("cluster", cs),
+				zap.Error(rerr),
+			)
+		} else {
+			d.lg.Info(
+				"discovery joined cluster successfully",
+				zap.String("cluster", cs),
+			)
+		}
+	}()
+
+	return d.joinCluster(config)
+}
+
+type discovery struct {
+	lg      *zap.Logger
+	cluster string
+	id      types.ID
+	c       *clientv3.Client
+	retries uint
```

Comment:
The retries seems to be accumulating between public API calls. 
It's not very intuitive. Either we should document that, or reset the counter on each public call.

Reply:
Yes, you are right. I reset it to 0 on successful request.